README.md
# Twisted Frequency 2024 Projection Source

## [Koura stage](koura/)

---

## Contributions

- Make an example project in your favourite visuals program
- Make some loops and distribute them on google drive
- Give me feedback at [vis@edt.nz](mailto:vis@edt.nz) if you find problems or have suggestions

### Rules

Check with the file owner before making a change to an existing file.


koura/CHANGELOG.md
# 3

- Added speakers
- Fixed pixel misalignment
- Separated fbx stage and scene to different files

# 2

- Separated working files into experiments, loops & stills
- Fixed sorting of uv faces to work better with baking
- Smaller mini koura
- Added heaps of separated stills for generators

# 1

Initial!


koura/loops/blender_bake_animation.py
import bpy
import os
from datetime import datetime

def print(data):
    for window in bpy.context.window_manager.windows:
        screen = window.screen
        for area in screen.areas:
            if area.type == 'CONSOLE':
                override = {'window': window, 'screen': screen, 'area': area}
                bpy.ops.console.scrollback_append(override, text=str(data), type="OUTPUT")

class BakeAnimationPanel(bpy.types.Panel):
    """Custom panel for baking animation using Eevee"""
    bl_idname = "OBJECT_PT_bake_animation"
    bl_label = "Bake Animation"
    bl_space_type = 'PROPERTIES'
    bl_region_type = 'WINDOW'
    bl_context = "object"

    def draw(self, context):
        layout = self.layout
        obj = context.object

        # Check if object is valid
        if not obj or obj.type not in {'MESH', 'CURVE'}:
            return

        # Add controls
        layout.prop(obj, "use_eevee_bake", text="Use Eevee")
        row = layout.row()
        row.prop(context.scene, "frame_start", text="Start Frame")
        row.prop(context.scene, "frame_end", text="End Frame")
        layout.prop(context.scene, "render.filepath", text="Output Directory")

        # Progress bar
        progress_bar = layout.prop(context.scene, "bake_progress", text="Progress")

        # Bake button
        layout.operator("bake_animation.operator", text="Bake")

class BakeAnimationOperator(bpy.types.Operator):
    """Operator for baking animation using Eevee with UI update"""
    bl_idname = "bake_animation.operator"
    bl_label = "Bake Animation"

    def execute(self, context):
        scene = bpy.context.scene
        obj = bpy.context.active_object
        
        image_name = obj.name + '_BakedTexture'
        img = bpy.data.images.new(image_name,3840,2160)

        # Set the frame range and frame step
        start_frame = scene.frame_start
        end_frame = scene.frame_end
        frame_step = 1
        
        filepath = bpy.data.filepath
        directory = os.path.dirname(filepath)
        
        now = datetime.now()
        now_str = now.strftime("%d-%m-%Y %H-%M-%S")
        
        relative_folder=f"outputs/{now_str}"
        folder = f"{directory}/{relative_folder}"
        if not os.path.exists(folder):
            os.makedirs(folder)
        print(f"saving files to {folder}")
        
        #Due to the presence of any multiple materials, it seems necessary to iterate on all the materials, and assign them a node + the image to bake.
        for mat in obj.data.materials:
            mat.use_nodes = True #Here it is assumed that the materials have been created with nodes, otherwise it would not be possible to assign a node for the Bake, so this step is a bit useless
            nodes = mat.node_tree.nodes
            texture_node = nodes.new('ShaderNodeTexImage')
            texture_node.name = 'Bake_node'
            texture_node.select = True
            nodes.active = texture_node
            texture_node.image = img #Assign the image to the node
        
        
        bpy.ops.wm.redraw_timer(type='DRAW_WIN_SWAP', iterations=1)

        # Loop through each frame
        for frame in range(start_frame, end_frame + 1, frame_step):
            # Set the current frame
            scene.frame_current = frame
            
            scene.bake_progress = frame
            
            # Bake the frame
#            bpy.ops.render.render(write_still=True, animation=False)
            bpy.context.view_layer.objects.active = obj
            bpy.ops.object.bake(type='COMBINED', save_mode="EXTERNAL", use_selected_to_active=False)

            # Save the baked image
#            baked_image = bpy.data.images['Render Result']
            path = f"{folder}/{obj.name}_{frame:04}.png"
            img.save_render(path)
            
            # Update UI after each frame
            bpy.ops.wm.redraw_timer(type='DRAW_WIN_SWAP', iterations=1)

        scene.bake_progress = 0
        
        #In the last step, we are going to delete the nodes we created earlier
        for mat in obj.data.materials:
            for n in mat.node_tree.nodes:
                if n.name == 'Bake_node':
                    mat.node_tree.nodes.remove(n)

        print("Baking completed!")

        return {'FINISHED'}

# Register custom properties
def register():
    bpy.utils.register_class(BakeAnimationPanel)
    bpy.utils.register_class(BakeAnimationOperator)
    bpy.types.Scene.bake_progress = bpy.props.FloatProperty(name="Bake Progress", default=0.0)

# Unregister custom properties
def unregister():
    bpy.utils.unregister_class(BakeAnimationPanel)
    bpy.utils.unregister_class(BakeAnimationOperator)
    del bpy.types.Scene.bake_progress

if __name__ == "__main__":
    register()


koura/loops/blender_bake_animation_from_camera_headless.py
import bpy
import os
from datetime import datetime
import sys
import bpy


def print(data):
    sys.stdout.write(str(data) + "\n")
    for window in bpy.context.window_manager.windows:
        screen = window.screen
        for area in screen.areas:
            if area.type == "CONSOLE":
                override = {"window": window, "screen": screen, "area": area}
                bpy.ops.console.scrollback_append(
                    override, text=str(data), type="OUTPUT"
                )


render_collection_name = "Gen"
reprojection_collection_name = "Reprojection"


def execute():
    scene = bpy.context.scene

    # obj = bpy.data.objects[object_name]
    render_collection = bpy.data.collections[render_collection_name]
    reprojection_collection = bpy.data.collections[reprojection_collection_name]

    reprojection_collection.hide_render = True
    # reprojection_collection.hide_viewport = False

    image_name = render_collection_name + "_BakedTexture"
    img = bpy.data.images.new(image_name, 3840, 2160, alpha=True)
    img.colorspace_settings.name = "AgX Base sRGB"

    # Set the frame range and frame step
    start_frame = scene.frame_start
    end_frame = scene.frame_end
    frame_step = 1

    filepath = bpy.data.filepath
    directory = os.path.dirname(filepath)

    now = datetime.now()
    now_str = now.strftime("%d-%m-%Y %H-%M-%S")

    relative_folder = f"outputs/{now_str}"
    folder = f"{directory}/{relative_folder}"
    if not os.path.exists(folder):
        os.makedirs(folder)

    print(f"Baking files to {folder}")

    BAKE_NODE_NAME = "Bake_node"
    RENDER_TEXTURE_NAME = "Render Result"
    RENDER_TEXTURE_NODE_NAME = "Image Texture"
    BAKE_TYPE = "COMBINED"
    # BAKE_PASS_FILTER = {"DIRECT", "INDIRECT"}

    # Due to the presence of any multiple materials, it seems necessary to iterate on all the materials, and assign them a node + the image to bake.
    for obj in reprojection_collection.objects:
        mat = obj.active_material
        mat.use_nodes = True  # Here it is assumed that the materials have been created with nodes, otherwise it would not be possible to assign a node for the Bake, so this step is a bit useless
        nodes = mat.node_tree.nodes
        texture_node = nodes.new("ShaderNodeTexImage")
        texture_node.name = BAKE_NODE_NAME
        texture_node.select = True
        nodes.active = texture_node
        texture_node.image = img  # Assign the image to the node

    # Loop through each frame
    for frame in range(start_frame, end_frame + 1, frame_step):
        # Set the current frame
        scene.frame_current = frame

        render_collection.hide_render = False
        # Set the render file location
        render_filepath = f"{folder}/render_{frame}.png"
        scene.render.filepath = render_filepath

        # Render the image
        bpy.ops.render.render(write_still=True)
        render_collection.hide_render = True

        # rendered_image = bpy.data.images[RENDER_TEXTURE_NAME]

        reprojection_collection.hide_render = False
        reprojection_collection.hide_viewport = False
        reprojection_collection.hide_select = False

        for obj in reprojection_collection.objects:
            mat = obj.active_material
            nodes = mat.node_tree.nodes
            texture_node = nodes[RENDER_TEXTURE_NODE_NAME]
            texture_node.image = bpy.data.images.load(render_filepath)
            nodes.active = nodes[BAKE_NODE_NAME]
            # obj.select_set(True)
            # texture_node.image.colorspace_settings.name = "AgX Base sRGB"
            #  = obj
            # obj.hide_set(False)
            print(bpy.context.view_layer.objects.active)
            bpy.ops.object.bake(
                type=BAKE_TYPE,
                # save_mode="EXTERNAL",
                use_selected_to_active=False,
                # pass_filter=BAKE_PASS_FILTER,
            )
            # img.save()

        reprojection_collection.hide_render = True
        reprojection_collection.hide_viewport = True
        reprojection_collection.hide_select = True

        # Save the baked image
        path = f"{folder}/{obj.name}_{frame:04}.png"
        img.save(filepath=path)

        print(f"Baked {frame}/{end_frame-start_frame}")

    # In the last step, we are going to delete the nodes we created earlier
    for mat in obj.data.materials:
        for n in mat.node_tree.nodes:
            if n.name == "Bake_node":
                mat.node_tree.nodes.remove(n)

    print("Baking completed!")

    return {"FINISHED"}


if __name__ == "__main__":
    execute()


koura/loops/blender_bake_animation_headless.py
import bpy
import os
from datetime import datetime
import sys


def print(data):
    sys.stdout.write(str(data) + "\n")
    for window in bpy.context.window_manager.windows:
        screen = window.screen
        for area in screen.areas:
            if area.type == "CONSOLE":
                override = {"window": window, "screen": screen, "area": area}
                bpy.ops.console.scrollback_append(
                    override, text=str(data), type="OUTPUT"
                )


object_name = "stage"


def execute():
    scene = bpy.context.scene
    obj = bpy.data.objects[object_name]

    image_name = obj.name + "_BakedTexture"
    img = bpy.data.images.new(image_name, 3840, 2160)

    # Set the frame range and frame step
    start_frame = scene.frame_start
    end_frame = scene.frame_end
    frame_step = 1

    filepath = bpy.data.filepath
    directory = os.path.dirname(filepath)

    now = datetime.now()
    now_str = now.strftime("%d-%m-%Y %H-%M-%S")

    relative_folder = f"outputs/{now_str}"
    folder = f"{directory}/{relative_folder}"
    if not os.path.exists(folder):
        os.makedirs(folder)

    print(f"Baking files to {folder}")

    bpy.context.view_layer.objects.active = obj

    # Due to the presence of any multiple materials, it seems necessary to iterate on all the materials, and assign them a node + the image to bake.
    for mat in obj.data.materials:
        mat.use_nodes = True  # Here it is assumed that the materials have been created with nodes, otherwise it would not be possible to assign a node for the Bake, so this step is a bit useless
        nodes = mat.node_tree.nodes
        texture_node = nodes.new("ShaderNodeTexImage")
        texture_node.name = "Bake_node"
        texture_node.select = True
        nodes.active = texture_node
        texture_node.image = img  # Assign the image to the node

    # Loop through each frame
    for frame in range(start_frame, end_frame + 1, frame_step):
        # Set the current frame
        scene.frame_current = frame

        # Bake the frame
        bpy.context.view_layer.objects.active = obj
        bpy.ops.object.bake(
            type="COMBINED", save_mode="EXTERNAL", use_selected_to_active=False
        )

        # Save the baked image
        path = f"{folder}/{obj.name}_{frame:04}.png"
        img.save_render(path)

        print(f"Baked {frame}/{end_frame-start_frame}")

    # In the last step, we are going to delete the nodes we created earlier
    for mat in obj.data.materials:
        for n in mat.node_tree.nodes:
            if n.name == "Bake_node":
                mat.node_tree.nodes.remove(n)

    print("Baking completed!")

    return {"FINISHED"}


if __name__ == "__main__":
    execute()
